### PR TITLE
Improve print to file

### PR DESCRIPTION
**Pull request by Vovan707**

### CODE REVIEW

**Comment by Vovan707** [comment]
@sergeyradkevich review please

**Comment by sergeyradkevich** [comment]
@Vovan707 As discussed in mail let's correct printing into a text file so it contains everything the menu outputs. Please do it within this pull request. Once ready, please post a comment here that it is updated.

**Comment by Vovan707** [comment]
OK, I realized

**Comment by Vovan707** [comment]
@sergeyradkevich  I update, review please

**Comment by sergeyradkevich** [comment]
Hey @Vovan707! I didn't have a chance to look at this. Will try to review tomorrow.

**Comment by sergeyradkevich** [comment]
Or today

File: Menu.java [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
 
 public class Menu {
 
 	Array arrayClass;
 	int[] arrays;
-	static int clics;
+	static int clics; 
 	private boolean status = true;
 
+	Print print = new Print();
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 It looks this field has no usages.

Comment (sergeyradkevich) [inline Menu.java]: Also is it from `net.print` package you've removed? If yes, then I'm wondering wethere this is the latest version of file? I mean it shouldn't  have compiled in this case.

File: Menu.java [hunk 2]
```diff
@@ -1,25 +1,33 @@
 package net.array_project;
+
+import java.io.*;
+
 import net.print.Print;
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 The same question is about this import.

Hey, do you recompile your source code before pusing the code to git?

Comment (sergeyradkevich) [inline Menu.java]: Do you rerun all tests before push to git?

File: Menu.java [hunk 3]
```diff
@@ -1,25 +1,33 @@
...
 
 public class Menu {
 
 	Array arrayClass;
 	int[] arrays;
-	static int clics;
+	static int clics; 
 	private boolean status = true;
 
+
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 This empty line is not needed here, please remove.

**Comment by sergeyradkevich** [comment]
@Vovan707 Please do not push the code unless a review is done by me. I will let you know.

File: Menu.java [hunk 4]
```diff
@@ -1,25 +1,33 @@
...
 	MenuItem avarageNumber = new MenuItem("3", "Show avarage number");
 	MenuItem showEvenNumber = new MenuItem("4", "Show even number");
 	MenuItem sortArray = new MenuItem("5", "Sort Array");
 	MenuItem exit = new MenuItem("q", "For exit");
-	MenuItem[] items = { randomFill, showArray, avarageNumber, showEvenNumber,sortArray, exit };
-			
+	MenuItem[] items = { randomFill, showArray, avarageNumber, showEvenNumber,
+			sortArray, exit };
+	
+
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 Please leave only one line to seperate fields from methods.

File: Menu.java [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
 	public Menu(int[] array) {
-		this.arrays = array;
+				this.arrays = array;
 		arrayClass = new Array(arrays);
 		clics = 0;
+		arrayClass.supportForOutput();
+
 	}
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 Look at formatting of this method. Indentation of the first line is poor and there is an empty line. Please reformat.

File: Menu.java [hunk 6]
```diff
@@ -27,13 +35,15 @@ public void setStatus(boolean status) {
...
+		System.out.println("\nPlease select function:\n");
 	}
 
 	public void display() {
 		banner();
 		for (MenuItem item : items) {
-			Print.println(item.displayName());
+			String s = item.displayName();
+			System.out.println(s);
+
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 The same here. No need to leave that blank line.

File: TestForPrint.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+import org.junit.Test;
+
+public class TestForPrint {
+
+	@Test(expected = FileNotFoundException.class)
+	public void testShowExeption() throws FileNotFoundException {
+		 Print.println("s");
+
+	}
```
Comment (sergeyradkevich) [inline TestForPrint.java]: @Vovan707 Such class even doesn't exist in your application. It looks you do not run all your tests.

**Comment by sergeyradkevich** [comment]
FYI: @Vovan707 What I've asked regarding posting of content of the file is to do the following:
```txt

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit

[4 1 0 7 2 7 6 4 1 9 ]

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit
	ARRAY: 

[4 1 0 7 2 7 6 4 1 9 ]

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit

[4 1 0 7 2 7 6 4 1 9 ]

Avarage from is 4.1

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit

[4 0 2 6 4 ]

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit
Sorted Array

[0 1 1 2 4 4 6 7 7 9 ]

Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit
You enter the wrong symbol !


Please select function:

  1 : Random fill
  2 : Show Array
  3 : Show avarage number
  4 : Show even number
  5 : Sort Array
  q : For exit
	Good bye! 
```

So we will remove the file from git later.

File: TestArray.java [hunk 8]
```diff
@@ -33,22 +43,34 @@ public void checkObject() {
...
+		 assertFalse(menuItem.equals(menuItem2));
 	}
 
 	@Test
-	public void equalTwoObject() {
-		assertFalse(menuItem.equals(menuItem2));
+	public void equalsTwoArray() {
+		
+		assertArrayEquals(arrayObj.getEvenArray(),new int[]{2});
+
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 These empty lines are killing me :)

File: TestArray.java [hunk 9]
```diff
@@ -33,22 +43,34 @@ public void checkObject() {
...
-	public void testDisplayName() {
+	public void checkAvarageNumber() {
+		double avarageNumber = arrayObj.getAvarage();
+		assertEquals(2.0, arrayObj.getAvarage());
+	}
 
-		assertTrue(menuItem.displayName(), true);
+	@Test
+	public void checkOrGetTrueConditionNumber() {
+		assertFalse(arrayObj.condition(1));
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 Please rename:
- the test into `nubmerIsNotEven`;
- rename method into `isEven`.

File: Array.java [hunk 10]
```diff
@@ -13,13 +17,19 @@ public Array(int arr[]) {
...
+	public void supportForOutput() {
+		FileOutputStream f;
+		try {
+			
+			f = new FileOutputStream("D:\\File.txt");
+			System.setOut(new PrintStream(f));
+		} catch (FileNotFoundException e) {
+
+			e.printStackTrace();
+		}
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 For now please reformat `try` block. About handling exception we will talk later on.

File: Array.java [hunk 11]
```diff
@@ -13,13 +17,19 @@ public Array(int arr[]) {
...
-	}
 
-	public void messagePrint(String s) {
-		System.out.print(s);
+	public void supportForOutput() {
+		FileOutputStream f;
+		try {
+			
+			f = new FileOutputStream("D:\\File.txt");
+			System.setOut(new PrintStream(f));
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 Now about the main point. You write to the file, finally :) Congrats!

Comment (sergeyradkevich) [inline Array.java]: @Vovan707 But there are some issues with such implementation. Let me explain.

**Comment by Vovan707** [comment]
@sergeyradkevich  review please

File: Array.java [hunk 12]
```diff
@@ -1,25 +1,16 @@
...
+package net.array_project;
+import java.util.*;
 
 public class Array {
 	int array[];
-
 	int oddNumberArray[];
 	int arrayEvenNumbers[];
-	public int evenNumber;
+	int evenNumber;
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 What is the purpose of this field? Please check whether it makes a sense to keep it.

File: Array.java [hunk 13]
```diff
@@ -27,14 +18,12 @@ public void computeNumber() {
 			}
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 It looks this method has no usage. Please check and delete.

File: TestArray.java [hunk 14]
```diff
@@ -1,55 +1,64 @@
...
+		assertEquals(2.0, arrayObj.getAvarage());
 	}
-
 	@Test
-	public void equalsTwoName() {
-		assertEquals(menuItem.displayName().equals(menuItem.getName()), false);
-
+	public void checkOrGetTrueConditionNumber() {
+		assertFalse(arrayObj.condition(1));
+	}
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 The following suggestion hasn't been implemented:
> @Vovan707 Please rename:
- the test into `nubmerIsNotEven`;
- rename method into `isEven`.

File: Menu.java [hunk 15]
```diff
@@ -15,68 +13,67 @@
...
+		}
 	}
-
 	public void setStatus(boolean status) {
 		this.status = status;
 	}
-
 	public void banner() {
-		Print.println("\nPlease select function:\n");
+		ps.println("\nPlease select function:\n");
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 In this context another developer would not be sure what `ps` means, can you please rename it into `print`, `output` or `display` or anything you think would represent the idea.

File: Menu.java [hunk 16]
```diff
@@ -15,68 +13,67 @@
...
 	}
-
 	public void display() {
 		banner();
 		for (MenuItem item : items) {
-			Print.println(item.displayName());
+			String s = item.displayName();
+			ps.println(s);
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 Please rename `s` into `name`.

File: Menu.java [hunk 17]
```diff
@@ -15,68 +13,67 @@
...
 	public boolean isActive() {
 		return status;
 	}
-
 	public void viewArray(int array[]) {
-		Print.print("\n[");
+		ps.print("\n[");
 		for (int arrayIn : array) {
-			Print.print(arrayIn + " ");
+			System.out.print(arrayIn + " ");
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 It looks this `System.out` was missed.

File: TestArray.java [hunk 18]
```diff
@@ -1,55 +1,64 @@
...
 	@Before
 	public void createClassObject() {
 		menuItem = new MenuItem("1", "FillAray");
+		arrayObj = new Array(array);
+		try {
+			OutputStream outputStream = new FileOutputStream("D:\\Test.txt");
+			ps = new PrintStream(outputStream);
+		} catch (FileNotFoundException e) {	
+			e.printStackTrace();
+		}
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 I'm not following the purpose of this change. Why do we need to create a print stream in the tests?

**Comment by sergeyradkevich** [comment]
@Vovan707 Reviewed. Please work on given suggestions.

File: Array.java (hunk 12 above)
Comment (Vovan707) [inline Array.java]: Yes, this fields not required

File: Array.java (hunk 13 above)
Comment (Vovan707) [inline Array.java]: I delete it

File: Menu.java (hunk 15 above)
Comment (Vovan707) [inline Menu.java]: ok, I rename this

File: Menu.java (hunk 16 above)
Comment (Vovan707) [inline Menu.java]: Ok

File: Menu.java (hunk 17 above)
Comment (Vovan707) [inline Menu.java]: Thanks, I changed it

File: TestArray.java (hunk 18 above)
Comment (Vovan707) [inline TestArray.java]: I wanted to test the output but it failed ( and I forgot delete this

**Comment by Vovan707** [comment]
@sergeyradkevich  reviewed please

File: Menu.java [hunk 19]
```diff
@@ -15,68 +13,67 @@
 	MenuItem sortArray = new MenuItem("5", "Sort Array");
 	MenuItem exit = new MenuItem("q", "For exit");
 	MenuItem[] items = { randomFill, showArray, avarageNumber, showEvenNumber,sortArray, exit };
-			
+	PrintStream printToFile;
+
```
Comment (sergeyradkevich) [inline Menu.java]: @Vovan707 Let's rename into something general, like `print`.

File: MenuItem.java [hunk 20]
```diff
@@ -1,29 +1,26 @@
...
 	public void setId(String id) {
 		this.id = id;
 	}
-
 	public String getId() {
 		return id;
 	}
 	public String displayName() {
-		return String.format(" %s : %s",getId(),getName());
+		return String.format("  %s : %s",getId(),getName());
```
Comment (sergeyradkevich) [inline MenuItem.java]: @Vovan707 Please revert formatting back.

File: Array.java [hunk 21]
```diff
@@ -43,7 +24,7 @@ public boolean condition(int n) {
...
 		}
 		int arrayPosition = 0;
-		arrayEvenNumbers = new int[amount];
+		 arrayEvenNumbers = new int[amount];
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 Reformat please.

File: RealisedArray.java [hunk 22]
```diff
@@ -1,21 +1,19 @@
...
-
+		Scanner scanner;
 		int[] array = new int[10];
 		Menu menu = new Menu(array);
-		
 		String menuItem;
 
 		while (menu.isActive()) {
-			Scanner scanner = new Scanner(System.in);
+			scanner = new Scanner(System.in);
```
Comment (sergeyradkevich) [inline RealisedArray.java]: @Vovan707 Good point :+1: Let's move the scanner out of the loop.

File: TestArray.java [hunk 23]
```diff
@@ -1,55 +1,58 @@
...
+import static org.junit.Assert.*;
+import net.array_project.MenuItem;
+import org.junit.*;
 
 public class TestArray {
+	int array[] = { 1, 2, 3 };
+	Array arrayObj;
 	MenuItem menuItem;
 	MenuItem menuItem2;
+	PrintStream ps;
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 Please remove unused `ps` var.

File: TestArray.java [hunk 24]
```diff
@@ -1,55 +1,58 @@
...
 	}
-
 	@Test
 	public void checkIsNull() {
 		assertNull(menuItem2);
-
+	} 
+	@Test 
+	public void equalTwoObject() {
+		 assertFalse(menuItem.equals(menuItem2));
```
Comment (sergeyradkevich) [inline TestArray.java]: @Vovan707 Does this test make any sense?

File: Array.java [hunk 25]
```diff
@@ -34,8 +15,8 @@ public boolean condition(int n) {
 		}
 		return false;
 	}
```
Comment (sergeyradkevich) [inline Array.java]: @Vovan707 Adjust the method so it returns the expression:

```java
public boolean isEven(int number) {
  return (number % 2 == 0);
}
```

**Comment by Vovan707** [comment]
@sergeyradkevich  I update, reviewed please

**Pull request merged**